tests/sessions_tests/tests.py

```diff
@@ -353,6 +354,28 @@ def test_session_load_does_not_create_record(self):
         # provided unknown key was cycled, not reused
         self.assertNotEqual(session.session_key, 'someunknownkey')
 
+    def test_ticket_21608(self):
+        # Test for #21608 - Logged out sessions are resurrected by concurrent requests.
+
+        # Create new session.
+        SESSION_KEY = 'tkniiqcv0js5reo8y0ofrywrhmwnt9xw'
+        s1 = self.backend(SESSION_KEY)
+        s1.save(must_create=True)
+
+        # Logout in an other context.
+        s2 = self.backend(SESSION_KEY)
+        s2.delete()
+
+        # Modify session in first context.
+        s1.modified = True
+        try:
+            s1.save()  # This should throw an exception as the session is deleted, not resurrect it.
```

Comment:
I made this change and that revealed that the bug is also present in the other backends. The root of the problem is that  save is implemented as an "upsert" and needs to just be a update as the state "logged out" is represented as the absence of a stored session. The only time a session should be created is at login as the presence of a stored session means that the user is logged in